### ISSUE TITLE
LineMaterialRealSampler does not sample along the given line

### ISSUE THREAD

**Issue by dschwen** [idaholab/moose#7546]
### Description of the enhancement or error report

`LineMaterialRealSampler` behaves differently from VPPs derived from `PointSamplerBase`. With `PointSamplerBase` the user obtains values along the specified line. `LineMaterialRealSampler` on the other hand does _not_ return points along the line, but quadrature points in elements intersected by the line. This is unexpected and should not be what such a general class does.

### Rationale for the enhancement or information for reproducing the error

I talked to @bwspenc and it seems that changing this over to use `reinitElemPhys` and evaluating the material properties at the points on the line might work for him as well. I suggest changing the behavior accordingly.

Ben needs this to sample a material property along a sample crossection and to fit a polynomial on the resulting data. I need it to output a radial energy distribution. Having proper evaluation across my elements would return a smoother results.

### Identified impact
Regolding will be required for the tests of this object.

**Comment by bwspenc** [comment]
I'm fine with renaming the existing class, but I think that the existing behavior is something that's still useful to have around.

**Comment by dschwen** [comment]
Can we move them into SolidMechanics then?

**Comment by bwspenc** [comment]
How about TensorMechanics?

**Comment by friedmud** [comment]
@dschwen I like the current behavior more than what you're proposing here.

The purpose of this PP is to give you back _exactly_ what the Kernel is seeing... for diagnostic purposes.

I'm ok with renaming this - but it should stay in MOOSE.

**Comment by friedmud** [comment]
Also: note that `reinitElemPhys()` won't ever allow you to sample stateful material properties... but this one does.

**Comment by dschwen** [comment]
Ok, that's fine with me, but we don't do that for `LineValueSampler` either and that is bugging me. Also, I'm not looking for a debugging tool, I want to output an energy density curve. The stateful materials are a good point to keep in mind though.

**Comment by friedmud** [comment]
Don't do what for `LineValueSampler`?  We have the ability to sample variables everywhere... so we don't.  Materials are only computed at discrete points... so that's what we've got here...

**Comment by dschwen** [comment]
We have the ability to sample material properties everywhere. Variables are only evaluated at quadrature points (in Kernels). I really don't see the difference here (modulo stateful material properties).

**Comment by friedmud** [comment]
Let me try again for you:

```
We have the ability to sample a subset of material properties everywhere.  However, the ones everyone wants to look at over time (because that's where most of the issues are) are the stateful ones which you can only get at quadrature points.  So the current object makes sense.
```

I think I captured what you were going for ;-)

**Comment by dschwen** [comment]
> the ones everyone wants to look at over time

_the ones @friedmud thinks everyone wants to look at over time_

Fixed that for you.

**Comment by dschwen** [comment]
Ah well, I'll just close this and write my own VPP in Marmot.

**Comment by friedmud** [comment]
Nah - I'm totally down with a new VPP that does what you're suggesting here.  I was just arguing for the merit of the existing object as well.  They can both exist in MOOSE (it might even be the same object... just with an option to either report along the line or report from the qps).

I just didn't want to see the existing capability go away!

**Comment by dschwen** [comment]
Ok, looking closer at the `LineValueSampler` hierarchy it seems that _variables_ are baked in pretty high up in the inheritance chain. I suggest adding a virtual method somewhere that performs the actual _get value_ at the current qp (where the qp is the point the element is initialized at). This would return `_coupled_moose_vars[j]->sln()[_qp];` (with `_qp == 0`) for `PointSamplerBase`.

Now we have two traits of the sample that need to be determined:

1. The shape of the point arrangement (Line..., Circle...)
2. The quantity that is sampled (...ValueSampler, ...FunctionSample, MaterialPropertySampler)

My first suggestion here would be templated inheritance (the way we discussed with the Terminator issue at the Tiger Team).

**Comment by friedmud** [comment]
@dschwen this is decent place to start introducing the ideas of templated inheritance.  I would like to do that slowly so we can grow best practices before going crazy with it.

If you want to take a shot at it here, I think it makes sense.
